hazelcast/src/main/java/com/hazelcast/map/impl/query/MapQueryEngineImpl.java

```diff
@@ -16,615 +16,158 @@
         }
-        return getSortedSubList(resultList, pagingPredicate, nearestAnchorEntry);
-    }
 
-    private boolean shouldUseCachedValue(MapContainer mapContainer) {
-        CacheDeserializedValues cacheDeserializedValues = mapContainer.getMapConfig().getCacheDeserializedValues();
-        switch (cacheDeserializedValues) {
-            case NEVER:
-                return false;
-            case ALWAYS:
-                return true;
-            default:
-                //if index exists then cached value is already set -> let's use it
-                return mapContainer.getIndexes().hasIndex();
-        }
-    }
-
-    @Override
-    public QueryResult queryLocalPartition(String mapName, Predicate predicate, int partitionId, IterationType iterationType) {
-        Collection<QueryableEntry> queryableEntries = queryTheLocalPartition(mapName, predicate, partitionId);
-        QueryResult result = newQueryResult(1, iterationType);
-        result.addAll(queryableEntries);
-        result.setPartitionIds(singletonList(partitionId));
         return result;
     }
 
+    // query thread first, fallback to partition thread
     @Override
-    public QueryResult invokeQueryLocalPartitions(String mapName, Predicate predicate, IterationType iterationType) {
-        checkNotPagingPredicate(predicate);
-
-        List<Integer> partitionIds = getLocalPartitionIds();
-        QueryResult result = newQueryResult(partitionIds.size(), iterationType);
+    public QueryResult runQueryOnAllPartitions(
+            String mapName, Predicate predicate, IterationType iterationType) {
+        Collection<Integer> mutablePartitionIds = getAllPartitionIds();
 
-        try {
-            Future<QueryResult> future = queryOnLocalMember(mapName, predicate, iterationType);
-            List<Future<QueryResult>> futures = singletonList(future);
-            addResultsOfPredicate(futures, result, partitionIds);
-            if (partitionIds.isEmpty()) {
-                return result;
-            }
-        } catch (Throwable t) {
-            if (t.getCause() instanceof QueryResultSizeExceededException) {
-                throw rethrow(t);
-            }
-            logger.warning("Could not get results", t);
-        }
-
-        try {
-            List<Future<QueryResult>> futures = queryPartitions(mapName, predicate, partitionIds, iterationType);
-            addResultsOfPredicate(futures, result, partitionIds);
-        } catch (Throwable t) {
-            throw rethrow(t);
+        QueryResult result = doRunQueryOnQueryThreads(
+                mapName, predicate, iterationType, mutablePartitionIds, ALL_MEMBERS);
+        if (isResultFromAnyPartitionMissing(mutablePartitionIds)) {
+            doRunQueryOnPartitionThreads(mapName, predicate, iterationType, mutablePartitionIds, result);
         }
 
         return result;
     }
 
+    // partition thread ONLY (for now)
```

Comment:
It means that this operation runs queries on partition threads only, bypassing the query threads.